include/swift/AST/Module.h

```diff
@@ -1073,23 +1073,22 @@ class SourceFile final : public FileUnit {
   }
 
   std::vector<Token> &getTokenVector() {
-    assert(EnabledAndAllCorrectedTokens.first && "Disabled");
-    return EnabledAndAllCorrectedTokens.second;
+    assert(pAllCorrectedTokens && "Disabled");
+    return *pAllCorrectedTokens;
   }
 
   ArrayRef<Token> getAllTokens() const {
-    assert(EnabledAndAllCorrectedTokens.first && "Disabled");
-    return EnabledAndAllCorrectedTokens.second;
+    assert(pAllCorrectedTokens && "Disabled");
+    return *pAllCorrectedTokens;
   }
 
   bool shouldKeepTokens() const {
-    return EnabledAndAllCorrectedTokens.first;
+    return (bool)pAllCorrectedTokens;
   }
 
 private:
-  /// Whether the SourceFile instance opts to collect underlying tokens and
-  /// the vector containing these tokens if so.
-  std::pair<bool, std::vector<Token>> EnabledAndAllCorrectedTokens;
+  /// If not null, the pointee vector should contain tokens of this source file.
+  std::unique_ptr<std::vector<Token>> pAllCorrectedTokens;
```

Comment:
Do you need the extra level of indirection, or can you just use Optional here?

Reply:
Yes, we do need the extra level. Is Optional any better than the unique pointer?

Reply:
That's what I meant by indirection, sorry. `std::unique_ptr` creates a separate allocation, while `Optional` reserves storage inline. I guess I don't actually know how big a `std::vector` is but I don't think it's going to make SourceFile unnecessarily large, and less malloc traffic is usually a good thing.